program/rust/src/validator.rs

```diff
@@ -0,0 +1,130 @@
+use {
+    crate::{
+        accounts::{
+            AccountHeader,
+            PriceAccount,
+            PriceAccountFlags,
+            PythAccount,
+            PythOracleSerialize,
+        },
+        c_oracle_header::PC_MAGIC,
+        error::OracleError,
+        processor::{
+            c_upd_aggregate,
+            c_upd_twap,
+        },
+        utils::pyth_assert,
+    },
+    solana_sdk::pubkey::Pubkey,
+    std::mem::size_of,
+};
+
+// Attempts to validate and access the contents of an account as a PriceAccount.
+fn validate_price_account(
+    price_account_info: &mut [u8],
+) -> Result<&mut PriceAccount, AggregationError> {
+    // TODO: don't return error on non-price account?
+    pyth_assert(
+        price_account_info.len() >= PriceAccount::MINIMUM_SIZE,
+        OracleError::AccountTooSmall.into(),
+    )
+    .map_err(|_| AggregationError::NotPriceFeedAccount)?;
+
+    {
+        let account_header = bytemuck::from_bytes::<AccountHeader>(
+            &price_account_info[0..size_of::<AccountHeader>()],
+        );
+
+        pyth_assert(
+            account_header.magic_number == PC_MAGIC
+                && account_header.account_type == PriceAccount::ACCOUNT_TYPE,
+            OracleError::InvalidAccountHeader.into(),
+        )
+        .map_err(|_| AggregationError::NotPriceFeedAccount)?;
+    }
+
+    let data = bytemuck::from_bytes_mut::<PriceAccount>(
+        &mut price_account_info[0..size_of::<PriceAccount>()],
+    );
+    if !data.flags.contains(PriceAccountFlags::ACCUMULATOR_V2) {
+        return Err(AggregationError::V1AggregationMode);
+    }
+    if !data
+        .flags
+        .contains(PriceAccountFlags::MESSAGE_BUFFER_CLEARED)
+    {
+        // We make sure that we don't generate v2 messages while v1 messages are still
+        // in the message buffer.
+        return Err(AggregationError::V1AggregationMode);
+    }
+
+    Ok(data)
+}
+
+fn update_aggregate(slot: u64, timestamp: i64, price_account: &mut PriceAccount) {
+    // NOTE: c_upd_aggregate must use a raw pointer to price data. We already
+    // have the exclusive mut reference so we can simply cast before calling
+    // the function.
```

Comment:
imo you need `if clock.slot > latest_aggregate_price.pub_slot_` here to avoid double aggregation on the slot were the flag gets turned on